feat(cli): add location data dir to deno info

crowlKats opened the pull request:
closes #10585

kt3k commented:
The implementation and tests look good to me 👍 

I'm a little worried about the word 'Location data'. Was that documented in our manual? What about 'Web Storage data location' as it only includes web storage data for now

crowlKats commented:
@kt3k maybe 'origin data' would be better?

crowlKats commented:
I changed it to `origin data` as it is a more fitting and less misleading name. Ideally the `location_data` dir in `DENO_DIR` should be renamed to `origin_data` as well, but since it would be breaking, i added a comment for 2.0

nayeemrmn commented on cli/flags.rs:
```diff
@@ -791,6 +791,7 @@ TypeScript compiler cache: Subdirectory containing TS compiler output.",
     .arg(Arg::with_name("file").takes_value(true).required(false))
     .arg(reload_arg().requires("file"))
     .arg(ca_file_arg())
+    .arg(location_arg())
```
```suggestion
    .arg(location_arg()).conflicts_with("file")
```

nayeemrmn commented on cli/tests/041_info_flag.out:
```diff
@@ -2,3 +2,4 @@ DENO_DIR location: "[WILDCARD]"
 Remote modules cache: "[WILDCARD]deps"
 Emitted modules cache: "[WILDCARD]gen"
 Language server registries cache: "[WILDCARD]registries"
+Origin data location: "[WILDCARD]location_data"
```
This should be individual paths to each DB file IMO (assuming there'll be another for indexeddb in the future)
```suggestion
Origin local storage: "[WILDCARD]local_storage"
```

crowlKats replied:
`local_storage` is a file in `location_data` under a hashed `--location`. indexeddb will be another file in the same dir, and same for other APIs that will require to have files on the system based on the origin (ie broadcast channel IPC)

nayeemrmn replied:
Yes it was a mistake to highlight this test. What will the output for `deno info --location https://foo/` be? I think it should be a direct path to the local storage db file, if it isn't already:
```
Local storage: /home/nayeem/.cache/deno/location_data/<hash>/local_storage
IndexedDB: /home/nayeem/.cache/deno/location_data/<hash>/indexed_db
```

crowlKats replied:
Currently it is
```
$  deno info --location=https://deno.land
Origin data location: "/Users/crowlkats/Library/Caches/deno/location_data/491d1c5f8d5eaa49d0ee3166c5aca7fbf9af5c466365713577ab3c739342712c"
```
I dont think we should have an entry for each API that will need to use persistent data based on the origin

nayeemrmn replied:
It's more like what we do with the other cache files. We should be as specific as possible and only stop where the structure within a directory is free-form, otherwise we are withholding info. It's just as unneeded as the 2nd and 3rd fields in:
```
DENO_DIR location: "/home/nayeem/.cache/deno"
Remote modules cache: "/home/nayeem/.cache/deno/deps"
Emitted modules cache: "/home/nayeem/.cache/deno/gen"
Language server registries cache: "/home/nayeem/.cache/deno/registries"
```

nayeemrmn replied:
Actually in that vein we should have the following fields:
```
Web storage: /home/nayeem/.cache/deno/location_data/<hash>
Local storage: /home/nayeem/.cache/deno/location_data/<hash>/local_storage
IndexedDB: /home/nayeem/.cache/deno/location_data/<hash>/indexed_db
```
It allows for more concise and less assuming consumption.

crowlKats replied:
> `Web storage: /home/nayeem/.cache/deno/location_data/<hash>`

why `Web storage`?

nayeemrmn replied:
The field is there same as `DENO_DIR location` for easy deletion. Or the name? `Origin data` would be more accurate for the broadcast channel stuff, that's fine.

crowlKats replied:
@nayeemrmn PTAL

nayeemrmn commented on cli/tests/041_info_flag_location.out:
```diff
@@ -0,0 +1,6 @@
+DENO_DIR location: "[WILDCARD]"
+Remote modules cache: "[WILDCARD]deps"
+Emitted modules cache: "[WILDCARD]gen"
+Language server registries cache: "[WILDCARD]registries"
```
The location variant shouldn't include the global fields, for modularity's sake.
```suggestion
```

crowlKats replied:
Done

lucacasonato replied:
I disagree with this. I don't understand why this would be more modular than including the other fields. Just seems like you need to run `deno info` multiple times if you want all info, which seems annoying.

nayeemrmn replied:
It's the same reason `deno info mod.ts` doesn't include the global fields. You'd run `deno info --location https://foo/` because you're interested in the local storage file for some specific origin, it's unlikely you'd want the global fields at the same time. `--location` shouldn't be considered a passive modifier of `deno info` since it would keep the other fields the same and just add new more specific ones. That makes `deno info --location <some_location>` more comparable to `deno info <some_module>` and it should be similarly disjoint.

nayeemrmn commented on cli/tests/041_info_flag_location.out:
```diff
@@ -0,0 +1,2 @@
+Origin data: "[WILDCARD]location_data[WILDCARD]"
```
I keep interpreting this as "origin info" as opposed to "per-origin storage bucket". How about:
```suggestion
Origin storage: "[WILDCARD]location_data"
```

crowlKats replied:
Done

nayeemrmn commented on cli/tests/041_info_flag.out:
```diff
@@ -2,3 +2,4 @@ DENO_DIR location: "[WILDCARD]"
 Remote modules cache: "[WILDCARD]deps"
 Emitted modules cache: "[WILDCARD]gen"
 Language server registries cache: "[WILDCARD]registries"
+Origin data: "[WILDCARD]location_data"
```
Ditto
```suggestion
Origin storage: "[WILDCARD]location_data"
```

nayeemrmn commented on cli/main.rs:
```diff
@@ -204,11 +204,12 @@ pub fn create_main_worker(
         .dir
         .root
         .clone()
+        // TODO(@crowlKats): change to origin_data for 2.0
         .join("location_data")
```
Do we need to consider this breaking? I feel like we provide cache paths in `deno info` so we can rename them while providing a stable handle, and this wasn't even in `deno info` yet.

crowlKats replied:
well renaming `location_data` would mean that users that have used local storage would lose their current data

lucacasonato replied:
I think we can still do this in 1.x - we add a "migration" that renames old to new. Lets do this in another PR though.

nayeemrmn commented on cli/flags.rs:
```diff
@@ -791,6 +791,7 @@ TypeScript compiler cache: Subdirectory containing TS compiler output.",
     .arg(Arg::with_name("file").takes_value(true).required(false))
     .arg(reload_arg().requires("file"))
     .arg(ca_file_arg())
+    .arg(location_arg().conflicts_with("file"))
```
This needs a `deno info`-specific help text. I think clap will allow overriding it here.

crowlKats replied:
What exactly should the help text contain?

nayeemrmn replied:
```
Show files used for web storage APIs etc. when running a script with '--location=<HREF>'
```

crowlKats replied:
Done

nayeemrmn reviewed (approved):
LGTM

lucacasonato commented on cli/flags.rs:
```diff
@@ -791,6 +791,11 @@ TypeScript compiler cache: Subdirectory containing TS compiler output.",
+    .arg(
+      location_arg()
+        .conflicts_with("file")
+        .help("Show files used for web storage APIs etc. when running a script with '--location=<HREF>'")
```
```suggestion
        .help("Show files used for origin bound APIs like the Web Storage API when running a script with '--location=<HREF>'")
```

lucacasonato commented on cli/main.rs:
```diff
@@ -266,37 +267,63 @@ where
-      colors::bold("Language server registries cache:"),
-      registry_cache,
-    );
+    if location.is_none() {
+      println!("{} {:?}", colors::bold("DENO_DIR location:"), deno_dir);
+      println!(
+        "{} {:?}",
+        colors::bold("Remote modules cache:"),
+        modules_cache
+      );
+      println!(
+        "{} {:?}",
+        colors::bold("Emitted modules cache:"),
+        typescript_cache
+      );
+      println!(
+        "{} {:?}",
+        colors::bold("Language server registries cache:"),
+        registry_cache,
+      );
+      println!("{} {:?}", colors::bold("Origin storage:"), origin_dir);
+    } else {
+      println!("{} {:?}", colors::bold("Origin storage:"), origin_dir);
```
```suggestion
    println!("{} {:?}", colors::bold("DENO_DIR location:"), deno_dir);
    println!(
      "{} {:?}",
      colors::bold("Remote modules cache:"),
      modules_cache
    );
    println!(
      "{} {:?}",
      colors::bold("Emitted modules cache:"),
      typescript_cache
    );
    println!(
      "{} {:?}",
      colors::bold("Language server registries cache:"),
      registry_cache,
    );
    println!("{} {:?}", colors::bold("Origin storage:"), origin_dir);
    if location.is_some() {
```

lucacasonato commented on cli/main.rs:
```diff
@@ -266,37 +267,63 @@ where
         "registryCache": registry_cache,
-    });
+        "originStorage": origin_dir,
+      })
```
I feel like location.is_some should still have all of the json output that is irrelevant to the location itself - so denoDir etc.

kt3k commented on cli/main.rs:
```diff
@@ -297,6 +312,14 @@ fn print_cache_info(
+    if location.is_some() {
+      println!(
+        "{} {:?}",
+        colors::bold("Local Storage:"),
```
nit: is this capitalization intentional?

crowlKats replied:
Yes, as "Local Storage" is a proper noun

kt3k reviewed (approved):
LGTM

kt3k commented:
@crowlKats Thank you for your contribution!
